clustering/__init__.py

```diff
@@ -10,3 +23,197 @@ def cluster(pd_rules: pd.DataFrame) -> dict:
+	#print(str(stability_score) , "%,")
+
+	return stability_score 
+
+
+def run_clustering():
+
+	
+	## Read in our environment variables for storage, tracking, and model ##
+
+	# Storage Parameters
+	ceph_key = os.environ.get("CEPH_KEY")
+	ceph_secret = os.environ.get("CEPH_SECRET")
+	ceph_host = os.environ.get("CEPH_HOST")
+	ceph_bucket = os.environ.get("CEPH_BUCKET")
+
+	# MLFlow Paramters
+	mlflow_experiment_id = os.environ.get('MLFLOW_EXPERIMENT_ID')
+	mlflow_tracking_uri  = os.environ.get("MLFLOW_TRACKING_URI")
+
+	
+	# Model Parameters
+	k_clusters = int(os.environ.get("K_CLUSTERS"))
+	pca_dimensions = int(os.environ.get("PCA_DIMENSIONS"))
+	date_1 = os.environ.get("DAY_1")
+	date_2 = os.environ.get("DAY_2")
+
+	# Parser to be used. This can be a configurable paramter in the futre, but for now we are focused only on this dataset. 
+	parser = "rule_data"
+
+
+	# Set up connection to our CEPH storage
+	client_kwargs = { 'endpoint_url' : ceph_host }
+	s3 = s3fs.S3FileSystem(secret=ceph_secret, key=ceph_key, client_kwargs=client_kwargs)
+
+	
+
+	## DAY 1 READ IN AND PREPROCESS ##
+
+	# Read in rules data for the first day 
+	url = "DH-DEV-INSIGHTS/{}/{}".format(date_1, parser)
+	rulesx_day_1 = pq.ParquetDataset(url,filesystem=s3).read_pandas().to_pandas()
+	print(len(rulesx_day_1),": data points from day 1")
+
+	# Preprocess the dataset
+	rules_day_1 = preprocess_rules_data(rulesx_day_1)
+
+
+	# normalize the data prior to PCA
+	min_max_scaler_1 = preprocessing.StandardScaler()
+	np_scaled_1 = min_max_scaler_1.fit_transform(rules_day_1)
+	data_1 = pd.DataFrame(np_scaled_1)
+
+	# Reduce the dimensionality of the data with PCA
+	pca_1 = PCA(n_components=pca_dimensions)
+	data_transformed_1 = pca_1.fit_transform(data_1)
+
+
+	## DAY 2 READ IN AND PREPROCESS ##
+
+	# Read in rules data for the first day 
+	url = "DH-DEV-INSIGHTS/{}/{}".format(date_2, parser)
+	rulesx_day_2 = pq.ParquetDataset(url,filesystem=s3).read_pandas().to_pandas()
+	print(len(rulesx_day_2),": data points from day 2")
+
+	# Preprocess the dataset
+	rules_day_2 = preprocess_rules_data(rulesx_day_2)
+
+
+	# normalize the data prior to PCA
+	min_max_scaler_2 = preprocessing.StandardScaler()
+	np_scaled_2 = min_max_scaler_2.fit_transform(rules_day_2)
+	data_2 = pd.DataFrame(np_scaled_2)
+
+	# Reduce the dimensionality of the data with PCA
+	pca_2 = PCA(n_components=pca_dimensions)
+	data_transformed_2 = pca_2.fit_transform(data_2)
+
+
+
+	## Perform the K-means clustering for both days and compare results ## 
+	
+	# probably over-kill here, but just to ensure we are tracking in the correct location, we re-define these variables. 
+	mlflow.set_tracking_uri("http://mlflow-server-svc-vpavlin-jupyterhub.cloud.paas.upshift.redhat.com")
+	os.environ["MLFLOW_EXPERIMENT_ID"]='2'
+	
+	# Start the mlflow tracking
+	with mlflow.start_run():
+		
+
+		# Run Kmeans on our dataset for day 1 and collect the lables
+		kmeans_1 = KMeans(n_clusters=k_clusters).fit(data_transformed_1)
```

Comment:
So creating kmeans_1 and kmeans_2, I believe the centroids start at random location? I.e. even if the clusters will be the same sizes and having the same systems, the positions will change. (so cluster kmeans_1 cluster "1" can be a cluster "3" in kmeans_2, etc. )

Reply:
@durandom so I am still in the dark of what we want to do. We want stable results, that means a systemX, should be always labeled 'label1', correct?

But now we don't even order the labels, so can't ensure that the 'label1' will keep having the same cluster center.

So shouldn't we train 1 model, figure out what the labels mean and then turn this to classification problem? 

---

We should probably start with: what we expect from this and how we'll interpret the result. And then we can talk about tests for the stability of the result?

Reply:
@Ladas yes. right, maybe classification is better for that. Let's discuss this once we have a PR going on for changing the clustering code itself.